spec/models/provider_spec.rb

```diff
@@ -0,0 +1,73 @@
+require 'rails_helper'
+
+RSpec.describe Provider, type: :model do
+
+  describe "Provider" do
+    it "is provider valid" do
+      expect(build(:provider)).to be_valid
+    end
+  end
+
+  describe "Validations" do
+    it { is_expected.to validate_uniqueness_of(:name) }
+  end
+
+  describe "Associations" do
+    it "has many warehouse_items" do
+      assc = described_class.reflect_on_association(:warehouse_items)
+      expect(assc.macro).to eq :has_many
+    end
+
+    it "has many warehouse_item_providers" do
+      assc = described_class.reflect_on_association(:warehouse_item_providers)
+      expect(assc.macro).to eq :has_many
+    end
+  end
+
+  describe "before_save" do
+     it { is_expected.to callback(:set_name).before(:save) }
+  end
+
+  describe "priority" do
+   it "enumerator value" do
+     expect(build(:provider, priority: 0).priority).to eq("contract")
+   end
+
+   it "enumerator value" do
+     expect(build(:provider, priority: 1).priority).to eq("individual_sale")
+   end
+    it { is_expected.to define_enum_for(:priority) }
+  end
+
+  describe "active" do
+    it "active true" do
+      create_list(:provider, 5, active: true)
+      create_list(:provider, 3, active: false)
+      expect(Provider.active.count).to eq(5)
+    end
+  end
+
+  describe "method :set_name" do
+    it "return name with method :set_name eq name_1" do
+      expect_name = create(:provider, name: "Amed")
+      expect_name1 = create(:provider, name: "Amed")
+      expect(exp_name1.name).to eq("Amed_1")
+    end
+
+    it "return name with method :set_name name" do
+      expect_name = create(:provider, name: "Amed")
+      expect(expect_name.name).to eq("Amed")
+    end
+  end
+
+  describe "method find_by_product_prioritized" do
+    it "return contract" do
```

Comment:
Debes describir lo que esperas que haga. En este caso esta regresando con una query en el metodo, una lista de registros del modelo Provider con ciertas condiciones. Entonces la descripcion seria algo asi como:

"returns a list of providers associated to product"

1.- tienes que hacer una en la que:

- crees una lista de productos (2 o 3 basta)
- de todos esos productos solo 1 no este asociado a algun proveedor
- al probar el metodo pases el id del producto que no tiene proveedor asociado
- debera regresar un array vacio ya que el id del producto al que le quieres buscar proveedor. No tiene proveedores asociados

2.- otra en la que:

- crees una lista de proveedores
- una lista de productos
- asocies todos los productos a algun proveedor
- pases un id de un producto que no existe. Puedes usar para sacar un id dinamico pero que no existe:
    -  Product.order(:id).last.id + 1 <--- esto va primero a ordenar los productos por el id, sacar el ultimo, sacar su id y agregarle +1. Recuerda que los ids son incrementales en la base de datos entonces si tu ultimo id es el 500, el siguiente que todavia no existe deberia ser el 501
- esta prueba debera esperar que regrese un array vacio ya que el id del producto ni siquiera existe

3.- Deberas hacer otra que pruebe el ordenamiento por prioridad. Los que tienen priority: 0 deben regresar primero en la lista de registros que sacas en la prueba. Entonces cuando crees tus proveedores, tendras que crear una lista de proveedores con priority 0, otra con priority 1 y asociarlos todos al mismo producto. deberan llegar ordenados por priority siendo 0 el de mayor prioridad osea el primero de la lista